src/dbnode/storage/index/types.go

```diff
@@ -79,13 +79,17 @@ type QueryResults struct {
 	Exhaustive bool
 }
 
-// Results is a collection of results for a query.
+// Results is a collection of results for a query, it is synchronized
+// when access to the results set is used as documented by the methods.
 type Results interface {
 	// Namespace returns the namespace associated with the result.
 	Namespace() ident.ID
 
-	// Map returns a map from seriesID -> seriesTags, comprising index results.
-	Map() *ResultsMap
+	// WithMap executes a function that has access to a results map
+	// from seriesID -> seriesTags, comprising index results.
+	// A function is required to ensure that the results is used
+	// while a read lock for the results is held.
+	WithMap(fn func(results *ResultsMap))
```

Comment:
instead of this, how would you feel about adding a `Seal()` and changing `Map()` to return `(*ResultsMap, error)`. Would allow callers to avoid the extra lambda.

Reply:
Yeah, I think this is actually the approach I'll take also to address the pooling issues that Richie was concerned about (i.e. not knowing when it's finalized and potentially adding values to it after it is finalized)

Reply:
@robskillington I think if you have an early time out / cancel, you have to give up and not return it to the pool cause even if you seal it there is no way to know when its safe to unseal it unless you ref count it

Reply:
So I now use a lifetime to protect against writing to the results once we return from the Query call to the index, this prevents writing to results during cancellation or any other early return code path.